### PR TITLE
refactor: initial work-in-progress implementation of streaming

### PR DESCRIPTION
**Pull request by connor4312**
This is the first PR to update some of the extension host side to
support streaming and 'infinitely large files' as needed for DAP.

Here I focus on creating a new "hex document model" which manages edits
and allows reading bytes with those edits applied. There's some gnarly
code here, so I also got a test setup going for some sanity checking.
The main additional complexity here is that this model now supports
insertions and removals in the middle of memory, not just at the end.
Though this probably won't be used in the debug case initially, I
thought it was still useful if I was rewriting things anyway...

Under the model lies the `fileAccessor` which is a simple I/O model. The
current accessor just uses the same VS Code APIs, but we could have some
extra casing that uses the native Node APIs to support ultra large
files when running in the Node extension host.

I also formalized the communication a bit for upcoming refactors there.
I haven't touched any of the webview-side code yet, and there are still
some parts of the extension host I haven't reworked yet (find/replace
mainly)

### CODE REVIEW

File: shared/hexDocumentModel.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+	 */
+	private readonly getSizeInner =
+		memoizeLast((_generation: number) => (this.readAccessor || this.accessor).getSize());
+
+	/**
+	 * Gets instructions for returning file data. Returns a list of contiguous
+	 * `Range` instances. Each "range" issues instructions until the next
+	 * "offset". This can be used to generate file data.
+	 */
+	private getEditTimeline() {
```
Comment (connor4312) [inline shared/hexDocumentModel.ts:252]: note: today we invalidate the entire timeline when an edit is made, but in theory we could just compute that onto the existing timeline. However I expect this to function to be fast enough as to make no difference

File: src/hexEditorProvider.ts [hunk 2]
```diff
@@ -269,77 +242,38 @@ export class HexEditorProvider implements vscode.CustomEditorProvider<HexDocumen
...
-							results: results
-						}
-					});
+					results = await document.searchProvider.createNewRequest().hexSearch(message.query);
 				}
+				return { type: MessageType.SearchResponse, results };
+			case MessageType.ReplaceRequest:
+				const edits = await document.replace(message.query, message.offsets, message.preserveCase);
+				return { type: MessageType.ReplaceResponse, edits };
 				return;
```
Comment (lramos15) [inline src/hexEditorProvider.ts:275]: Extra return

File: src/test/hexDocumentModel.test.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+			expect(await model.size()).to.equal(value.length);
+		};
+
+		it("reads file verbatim", async () => {
+			await assertContents(new Uint8Array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));
+		});
+
+		it("reads with a simple insert", async () => {
+			model.makeEdits([
```
Comment (lramos15) [inline src/test/hexDocumentModel.test.ts:38]: Some weird spacing going on in some of these documents. Needs formatting

**Review by lramos15 — CHANGES_REQUESTED** [review]
Great work, just some small things. Promising to see where this goes

File: shared/fileAccessor.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import type * as vscode from "vscode";
+
+export interface FileWriteOp {
+	offset: number;
+	data: Uint8Array;
+}
+
+export interface FileAccessor {
+	readonly uri: string;
```
[the remark this answers is not in the document]
Comment (connor4312) [reply shared/fileAccessor.ts:13]: This code is shared with the web view, where `vscode` isn't available

File: shared/fileAccessor.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+	/** Calculates the size of the associated document. Undefined if unbounded */
+	getSize(): Promise<number | undefined>;
```
[the remark this answers is not in the document]
Comment (connor4312) [reply shared/fileAccessor.ts:16]: My main reasoning is that this forces the unbound case to be dealt with explicitly -- a consumer can always `?? Infinity`, but they need to consciously do so

File: shared/hexDocumentModel.ts [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+	 */
+	public get edits(): readonly HexDocumentEdit[] {
+		return this._edits;
+	}
+
+	/**
+	 * Gets whether there are unsaved edits on the model.
+	 */
+	public get isDirty(): boolean {
+		return this._edits.length > 0;
```
[the remark this answers is not in the document]
Comment (connor4312) [reply shared/hexDocumentModel.ts:128]: Yea

File: src/hexDocument.ts [hunk 7]
```diff
@@ -53,27 +28,20 @@ export class HexDocument extends Disposable implements vscode.CustomDocument {
 			}
 		*/
 		telemetryReporter.sendTelemetryEvent("fileOpen", {}, { "fileSize": fileSize });
-		let fileData: Uint8Array;
 		const maxFileSize = (vscode.workspace.getConfiguration().get("hexeditor.maxFileSize") as number) * 1000000;
+		const isLargeFile = fileSize > maxFileSize && !backupId;
```
[the remark this answers is not in the document]
Comment (connor4312) [reply src/hexDocument.ts:32]: The file is never read here, it's only read upon request from the webview. This flag just signals to the webview to show a prompt before requesting any bytes.

File: src/test/hexDocumentModel.test.ts (hunk 3 above)
Comment (connor4312) [reply src/test/hexDocumentModel.test.ts:38]: This file is all tabs. But after things are merged I suggest adopting Prettier 🙂

**Pull request merged**